### DISCUSSION TITLE
WRIMS-core Comparison Design

### DISCUSSION THREAD

**Discussion by dezidizon** [CVWRSM/cvwrsm#63] [Ideas]
# Definitions
- Test Configuration - a collection of automated model runs and results analysis, selected to compare new model results with results from accepted models, either for regression testing in support of a new release or to demonstrate that new code produces the desired new result relative to an existing model. A file that specifies a particular test configuration is selected to launch a comparison.
- Model configuration - the set of files (.wresl, and others) that defines the network, inputs, operating rules, and so on for a collection of one or more WRIMS runs. These are the contents of the zip files that RMA gets from Hao Xie when we ask for a model to test.
- Run configuration - the instructions that control an execution of a single WRIMS run. Practically speaking, this is one file with a .config extension that is passed as an argument to the main function of a <code>wrimsv2.components.ControllerBatch</code> when a WRIMS compute is launched.
- Results set - Output files generated by a WRIMS run. Includes, but not limited to, DSS files. Results sets generated by programs built from new code with be compared to results sets generated from trusted models, which are referred to as "base results sets".
- Comparison configuration - specifies a run configuration to produce new results, a base result set to compare against, a comparison type, and settings appropriate for the comparison type.

# Test Configuration
1. Acquire and select model configurations for comparison tests
    1. A Model Configuration is made up of a directory that contains
        1. A RUN subdirectory that holds many additional subdirectories containing the model network data, mostly in .wresl files. The contents of the RUN folder do not change during a compute.
            1. There is some flexibility for using switching features of a model off or on within a model configuration. A main .wresl file identifies features used in a particular run, and a model configuration's RUN directory may contain more than one of these files, to be selected in a run configuration.
            2. The RUN directory also contains an EXTERNAL directory that holds java classes and DLLs that are used at runtime by WRIMS.
        2. A DSS subdirectory that holds both input and output data. Input DSS files are at the first level of the directory, and output DSS files are in an OUTPUT subdirectory. Only the contents of the DSS/OUTPUT folder are updated during a model run.
        3. Configuration files that control specific runs. When a model is run from the WRIMS GUI, the user's selections for run start and end times, names of input and output files, etc.  are stored in _study.config. The name of this file is passed to a ControllerBatch object to launch an execution from the GUI. By saving these files with unique names, we can set up batch processes to run specific WRIMS computes automatically, bypassing the GUI.     
      2. Currently available model configurations at RMA
        1. CalSim model configuration DCR21_6.68_BL_wsidi provided by Hao
            1. confirmed as runnable in the WRIMs build from CentralValleyModeling/Feature/HecDSS7 as built by Hao
            2. Full run from 1921 to 2015 takes about 2-1/2 hours
            3. A run for one year (1921-22) takes about 2 minutes
        2. CalLite model configuration CalLite4.1_TF_test provided by Hao
            1. Confirmed as runnable, same as above   
            2. Runs in under a minute for 1921-22
    3. More model configurations can be added to increase coverage of comparison tests   
2. Build base results sets
    1. Results sets form new WRIMS builds will be compared to pre-computed base results sets. We will not run the baseline models to generate reference results live for each test.
    2. Where DWR has provided output files from their own runs, we will use those as base results sets.
    3. Where existing output files are not available, we will run each model in a reference/baseline version of WRIMS/CalSim/CalLite/other to produce reference base results sets.
        1. A single model configuration can be used for runs of different lengths - one year v period-of-record, for example - or runs with different solvers.
        2. Create separate run configurations and results sets for each comparison we want, so that we have on run configuration file per set of reference outputs. 
3. Store model configurations and reference solution sets in zip files containing
    1. One model configuration (wresl files, input data, etc.)
    2. One or more run configuration files.
    3. For each run configuration file, a set of reference results. These will be stored in a directory named "BaseResults" and placed in sub-directories named to match the run configurations.
4. Store the model configurations and reference model results as artifacts in GitHub Packages and retrieve them when tests are run.

# Test Execution
1. A comparison configuration file will name the model configurations and run configurations to be used in a comparison.
    1. Suggest JSON or YAML for configuration file format. No custom parsers, please.
    2. The test configuration file contains a list of model configurations, and within each model configuration, a list of one or more run configurations, and a list of comparisons.
    3. In the comparison configuration file, a "model configuration" contains:
        1. The name of model configuration
        2. The zip file that contains the model configuration
        3. A list of run configurations to be tested. The run configuration file and all its reference outputs must be present in the zip file.
    4. In the comparison configuration file, a "comparison" contains:
        1. The name of the run configuration file to be executed
        2. The type of comparison to be carried out
            1. CalLite report
            2. EPPT comparison (defer)
            3. Regressoin test
            4. Others? (Excel, for one. Also defer)
2. Use a gradle javaExec task to run each configuration in the WRIMS engine with <code>wrimsv2.components.ControllerBatch</code>
3. Execute comparison tasks
    1. For a CalLite report or comparison
        1. The CalLite report tool uses the same template file as EPPT. Parts of that file are over-written in EPPT to apply it to the user's selected scenarios, file names are used without change in the CalLite tool.
        2. Templates will be configured to use fixed DSS file names as inputs, pointing to a base-case DSS results file and then the results DSS generated by the automated compute.
        3. The report tool will create a PDF report which can be stored as an artifact of the build.
        4. EPPT and the Callite tool both rely on Vista for DSS access, so neither can read DSS 7 output. For either tool to work for this purpose, revisions to code are required. The effort seems to be smaller for the CalLite tool, since there's not effort required to make it run as a batch process. There will be some effort required to maintain this as a separate utility, but it should be small.
    2. For a regression test, the following steps can be programmed as a JUnit test:
        1. The comparison configuration references a particular model run (i.e. a model configuration and one of the run configurations that depends on it). The run config file identifies the DV output file DSS file, which can be compared to the corresponding DSS file in the base results set for the run.
        2. Generate a DssCompare report (a DataSetResults object) with the above DSS files as inputs.
            1. Parameterize the comparison with path filters and corresponding tolerances.
            2. Test that <code>DataSetResults.pathsOutOfTolCounter</code> equals zero.
            3. Test for differences in text output as needed (implement later)
            4. Test for differences in non-DSS binary output as needed (implement later)
        3. Comparison reports could also be created using EPPT or Excel with the DSS add-in. At present, DWR is only asking for the CalLite Report Tool to be implemented, so these will be considered out-of-scope.
4. Applications:
    1. Automatically run a regression test with a short runtime, like the 1-year CalSim and CalLite tests run in the prototyping, for a push or a pull request.
    2. Manually run a more extensive configuraiton, including period-of-record runs, for a PR to main, when a tagged version is built, or on demand
5. Prototype Configuration file:
`model configurations:`
` - name: CalLite 01`
`   description: CalLite 4.1 example project`
`   zip file: watershed01`
`   runs in: testprojects/callite4.1_tf`
`   model runs:`
`     - name: CL test 01`
`       configuration:   cl_test01.config`
`       outfile: dss/output/test_01.dss`
`     - name: CL test 02`
`       configuration: cl_test02.config`
`       outfile: dss/output/test_02.dss`
` - name: CalSim 01`
`   description: CalSim example project`
`   zip file: watershed02`
`   model runs:`
`     - name: cs test 01`
`       configuration: cs_test01.config`
`       outfile: dss/output/test_01.dss`
`     - name: cs test 02`
`       configuraiton: cs_test02.config`
`       outfile: dss/output/test_02.dss`
`comparison configurations:`
` - name: callite update`
`   description: produces report of base v update results`
`   model configuraiton: callite_01`
`   run configuration: cl_test_01.config`
`   base output: BaseResults/cL_test_01/CL_DV_4.1_TF.DSS`
`   compare type: report`
`   report template: cl_update.inp`
`   report file: CalLite_Update.pdf`
` - name: CalLite Regression`
`   description: tests for deviation outside tolerance`
`   model configuration: CalLite 01`
`   run configuration: CL Test 01`
`   compare type: regression`
`   base output: BaseResults/cL_test_01/CL_DV_4.1_TF.DSS`
`   tolerances:`
`    - ["C = STOR*'', 0.05]`
`    - ["C = FLOW*'', 0.01]`
`    - ["C = FLOW F=2020AL", 0]`

6. Workflow corresponding to configuration file:
    1. User triggers gradle task to run a comparison report
        1. Launch options:
            1. Local: 'gradlew :wrims-comparison-test:runCompare - PcompareConfig=filename.yml'
            2. GitHub: run as workflow_dispatch with compare config filename as input
        2. Selection of configuration file determines which test runs are executed and the comparisons that will be run.
    2. As needed, wrims-core is built
    3. Gradle reads the configuration file and identifies the data sets that need to be downloaded
    4. Model configurations are downloaded from the artifact repository and extracted to wrims-comparison-test/build/testProjects/. The zipfile must contain a properly configured CalSim/CalLite project - with .config files to match those named in the test configuration file positioned parallel to the RUN folder.
    5. For each run configuration in the configuration file, a gradle JavaExec task is run to execute the WRIMS compute with the associated configuration file. Note that the configuration file specifies the name of the output DSS file. Since gradle will execute all the runs before generating the reports, it's important that each run configuration write its results to a uniquely named output DSS file.
    6. For each comparison configuration in the configuration file, a gradle JavaExec task is run to generate a report (for compare type = report) or a go/no-go response (for compare type = regression).
    7. The reports can be stored as artifacts. A plan or naming convention should be worked out so that the reports will be properly associated with the build being tested.
7. Implementation:
    1. Only implementing report type comparisons at present.
    2. Reports will be generated using an update to the CalLite Report Tool that will work iwth DSS 7 libraries
    3. Create a new repository for the CalLite Report Tool, because it's independent of other WRIMS development.
        1. Not part of WRIMS GUI
        2. We don't expect to require a new report tool with every pass through the WRIMS development cycle.
        3. Once dependencies on Vista are removed, it has no dependencies on other WRIMS code, just HEC monolith and external tools like jfreechart and iText that are used to make the report.

**Comment by rma-tom** [comment]
@hxiedwr @dwr-zroy @dwr-ycheng 
See comments in issue https://github.com/CentralValleyModeling/wrims/issues/134

We can continue discussing ideas about comparison tests here.
